File: templates/python/v1/flyweight/text_style_factory.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Tuple


@dataclass(frozen=True, slots=True)
class TextStyle:
    font_family: str
    font_size: int
    color_hex: str
    bold: bool
    italic: bool


@dataclass(slots=True)
class StyledText:
    text: str
    style: TextStyle

    def render(self) -> str:
        return (
            f"[{self.style.font_family},{self.style.font_size},{self.style.color_hex},"
            f"bold={self.style.bold},italic={self.style.italic}] {self.text}"
        )


class TextStyleFactory:
    def __init__(self) -> None:
        self._cache: Dict[Tuple[str, int, str, bool, bool], TextStyle] = {}

    def get_style(
        self,
        font_family: str,
        font_size: int,
        color_hex: str,
        bold: bool,
        italic: bool,
    ) -> TextStyle:
        key = (font_family, font_size, color_hex, bold, italic)
        if key not in self._cache:
            self._cache[key] = TextStyle(*key)
        return self._cache[key]

    def cache_size(self) -> int:
        return len(self._cache)

File: templates/javascript/v1/flyweight/textStyleFactory.js
const cache = new Map();
const KEY_SEPARATOR = "|";

function makeKey(fontFamily, fontSize, colorHex, bold, italic) {
  return [fontFamily, fontSize, colorHex, bold, italic].join(KEY_SEPARATOR);
}

function createStyle(fontFamily, fontSize, colorHex, bold, italic) {
  return Object.freeze({ fontFamily, fontSize, colorHex, bold, italic });
}

export function getTextStyle(fontFamily, fontSize, colorHex, bold, italic) {
  const key = makeKey(fontFamily, fontSize, colorHex, bold, italic);
  if (!cache.has(key)) {
    cache.set(key, createStyle(fontFamily, fontSize, colorHex, bold, italic));
  }

  return cache.get(key);
}

export function cacheSize() {
  return cache.size;
}

export class StyledText {
  constructor(text, style) {
    this.text = text;
    this.style = style;
  }

  render() {
    const { fontFamily, fontSize, colorHex, bold, italic } = this.style;
    return `[${fontFamily},${fontSize},${colorHex},bold=${bold},italic=${italic}] ${this.text}`;
  }
}
